core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1012,16 +1058,33 @@ class KafkaController(val config: KafkaConfig,
       }
     }
     controllerContext.clearPartitionLeadershipInfo()
+    info(s"Finished setting up partitionAssignments map ${timing(taskStartMs, failoverStartMs)}")
+
+    taskStartMs = time.milliseconds()
     controllerContext.shuttingDownBrokerIds.clear()
     controllerContext.shuttingDownBrokerIds ++= zkClient.getBrokerShutdownEntries
     // register broker modifications handlers
     registerBrokerModificationsHandler(controllerContext.liveOrShuttingDownBrokerIds)
-    // update the leader and isr cache for all existing partitions from Zookeeper
-    updateLeaderAndIsrCache()
+    info(s"Handled shutting-down brokers and registered update-handlers ${timing(taskStartMs, failoverStartMs)}")
+
+    // Update the leader and isr cache for all existing partitions from Zookeeper. THIS IS THE SLOWEST PART
+    // OF SEQUENTIAL CONTROLLER INIT (70% of overall time for 1500 topics, 15000 partitions in an integ test).
+    // [In parallel mode with 10 threads, partition-level init drops below 40% for the same test case.]
+    taskStartMs = time.milliseconds()
+    val allPartitions = controllerContext.allPartitions  // LUCAS: safe to cache this?  [given childChangeHandlers above, seems not entirely safe, yet allPartitions method itself makes a copy (toSet), so how does that work?]
```

Comment:
My answer to your question in the comment:
The controller is single-threaded. If some change happens on zk, the corresponding event, e.g. TopicChange, will be enqueued first, and processed after the controller fail-over process is done.
Thus making a copy of all partitions should be safe.